protocol/xmrmaker/message_handler.go

```diff
@@ -182,23 +183,40 @@ func (s *swapState) handleNotifyETHLocked(msg *message.NotifyETHLocked) (net.Mes
 		Address: string(addrAB),
 	}
 
+	go s.runT0ExpirationHandler()
+
+	s.setNextExpectedMessage(&message.NotifyReady{})
+	return out, nil
+}
+
+func (s *swapState) runT0ExpirationHandler() {
+	log.Debugf("time until t0 (%s): %vs",
+		s.t0.Format(common.TimeFmtSecs),
+		time.Until(s.t0).Seconds(),
+	)
+
+	waitCtx, waitCtxCancel := context.WithCancel(context.Background())
+	defer waitCtxCancel() // Unblock WaitForTimestamp if still running when we exit
+
+	waitCh := make(chan error)
 	go func() {
-		until := time.Until(s.t0)
-		log.Debugf("time until t0: %vs", until.Seconds())
+		waitCh <- s.WaitForTimestamp(waitCtx, s.t0)
+		close(waitCh)
+	}()
 
-		select {
-		case <-s.ctx.Done():
-			return
-		// TODO: Document why we add one second
-		case <-time.After(until + time.Second):
-			s.handleT0Expired()
-		case <-s.readyCh:
+	select {
+	case <-s.ctx.Done():
+		return
+	case <-s.readyCh:
+		return
+	case err := <-waitCh:
+		if err != nil {
+			// TODO: Do we propagate this error? If we retry, the logic should probably be inside WaitForTimestamp.
+			log.Errorf("Failure waiting for T0 timeout: err=%s", err)
```

Comment:
the only errors would be if the subscription fails for some reason (bad, should probably retry), or if the context is cancelled (which is fine)

do you know in what cases the subscription might error? i guess if the ethclient connection goes down?

Reply:
If you are using a local ethereum node (which users should), it is probably something that will never error.

I think the way to figure the answer out, will be to run some tests against a very-unreliable public node like wss://goerli-light.eth.linkpool.io/ws and see what errors happen in practice.